nibabel/brikhead.py

```diff
@@ -0,0 +1,558 @@
+    rw = False
+    ImageArrayProxy = AFNIArrayProxy
+
+    @classmethod
+    @kw_only_meth(1)
+    def from_file_map(klass, file_map, mmap=True):
+        """
+        Creates an AFNIImage instance from `file_map`
+
+        Parameters
+        ----------
+        file_map : dict
+            dict with keys ``image, header`` and values being fileholder
+            objects for the respective BRIK and HEAD files
+        mmap : {True, False, 'c', 'r'}, optional, keyword only
+            `mmap` controls the use of numpy memory mapping for reading image
+            array data.  If False, do not try numpy ``memmap`` for data array.
+            If one of {'c', 'r'}, try numpy memmap with ``mode=mmap``.  A
+            `mmap` value of True gives the same behavior as ``mmap='c'``.  If
+            image data file cannot be memory-mapped, ignore `mmap` value and
+            read array from file.
+        """
+        with file_map['header'].get_prepare_fileobj('rt') as hdr_fobj:
+            hdr = klass.header_class.from_fileobj(hdr_fobj)
+        imgf = file_map['image'].fileobj
+        if imgf is None:
+            imgf = file_map['image'].filename
+        data = klass.ImageArrayProxy(imgf, hdr.copy(), mmap=mmap)
+        return klass(data, hdr.get_affine(), header=hdr, extra=None,
+                     file_map=file_map)
+
+    @classmethod
+    @kw_only_meth(1)
+    def from_filename(klass, filename, mmap=True):
+        """
+        Creates an AFNIImage instance from `filename`
+
+        Parameters
+        ----------
+        filename : str
+            Path to BRIK or HEAD file to be loaded
+        mmap : {True, False, 'c', 'r'}, optional, keyword only
+            `mmap` controls the use of numpy memory mapping for reading image
+            array data.  If False, do not try numpy ``memmap`` for data array.
+            If one of {'c', 'r'}, try numpy memmap with ``mode=mmap``.  A
+            `mmap` value of True gives the same behavior as ``mmap='c'``.  If
+            image data file cannot be memory-mapped, ignore `mmap` value and
+            read array from file.
+        """
+        file_map = klass.filespec_to_file_map(filename)
+        return klass.from_file_map(file_map, mmap=mmap)
+
+    @classmethod
+    def filespec_to_file_map(klass, filespec):
+        """
+        Make `file_map` from filename `filespec`
+
+        Deals with idiosyncracies of AFNI BRIK / HEAD formats
+
+        Parameters
+        ----------
+        filespec : str
+            Filename that might be for this image file type.
+
+        Returns
+        -------
+        file_map : dict
+            dict with keys ``image`` and ``header`` where values are fileholder
+            objects for the respective BRIK and HEAD files
+
+        Raises
+        ------
+        ImageFileError
+            If `filespec` is not recognizable as being a filename for this
+            image type.
+        """
+        # copied from filebasedimages.py
```

Comment:
Can you post-process instead?  As in, inherit from `FileBasedImage` (sorry - why aren't you doing that? - I'm sure there's a good reason), and post-process the result?

```
file_map = super(AFNIImage, klass).filespec_to_file_map(filespec)
```

and then use `file_map['image'].filename` and `file_map['header'].filename`?

Do you agree that it seems cleaner to raise an error for `my_image.HEAD.gz`, even if `my_image.BRIK.gz` exists?